pkg/controller/uniteddeployment/uniteddeployment_controller.go

```diff
@@ -85,34 +104,198 @@ var _ reconcile.Reconciler = &ReconcileUnitedDeployment{}
+
+	nextReplicas := calcNextReplicas(nameToSubset, allocatedReplicas, nextPartitions, subsetType)
+	klog.V(4).Infof("Get UnitedDeployment %s/%s next replicas %v", instance.Namespace, instance.Name, nextReplicas)
+
+	if err := r.manageSubsets(instance, nameToSubset, nextReplicas, nextPartitions, currentRevision, updatedRevision, subsetType); err != nil {
+		klog.Errorf("Fail to update UnitedDeployment %s/%s: %s", instance.Namespace, instance.Name, err)
+		r.recorder.Event(instance.DeepCopy(), corev1.EventTypeWarning, fmt.Sprintf("Failed%s", eventTypeSubsetsUpdate), err.Error())
+		return reconcile.Result{}, nil
+	}
+
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileUnitedDeployment) getNameToSubset(instance *appsv1alpha1.UnitedDeployment, control subset.ControlInterface) (nameToSubset map[string]*subset.Subset, err error) {
+	subSets, err := control.GetAllSubsets(instance)
+	if err != nil {
+		r.recorder.Event(instance.DeepCopy(), corev1.EventTypeWarning, fmt.Sprintf("Failed%s", eventTypeFindSubsets), err.Error())
+		return nil, fmt.Errorf("fail to get all Subsets for UnitedDeployment %s/%s: %s", instance.Namespace, instance.Name, err)
+	}
+
+	klog.V(4).Infof("Classify UnitedDeployment %s/%s by subSet name", instance.Namespace, instance.Name)
+	nameToSubsets := r.classifySubsetByName(instance, subSets)
+
+	nameToSubset, err = r.deleteDupSubset(instance, nameToSubsets, control)
+	if err != nil {
+		r.recorder.Event(instance.DeepCopy(), corev1.EventTypeWarning, fmt.Sprintf("Failed%s", eventTypeDupSubsetsDelete), err.Error())
+		return nil, fmt.Errorf("fail to manage duplicate Subset of UnitedDeployment %s/%s: %s", instance.Namespace, instance.Name, err)
+	}
+
+	return
+}
+
+func getEffectiveSubsets(ud *appsv1alpha1.UnitedDeployment, nameToSubset map[string]*subset.Subset) (effectiveSubsets map[string]*subset.Subset) {
+	effectiveSubsets = map[string]*subset.Subset{}
+
+	for _, subsetDef := range ud.Spec.Topology.Subsets {
+		if ss, exist := nameToSubset[subsetDef.Name]; exist {
+			effectiveSubsets[subsetDef.Name] = ss
+		} else {
+			effectiveSubsets[subsetDef.Name] = &subset.Subset{}
+		}
+	}
+
+	return effectiveSubsets
+}
+
+func calcNextPartitions(ud *appsv1alpha1.UnitedDeployment, subsetReplicas map[string]int32) map[string]int32 {
+	partitions := map[string]int32{}
+	for _, subset := range ud.Spec.Topology.Subsets {
+		if partition, exist := ud.Spec.Strategy.Partitions[subset.Name]; exist {
+			partitions[subset.Name] = *partition
+		} else {
+			partitions[subset.Name] = subsetReplicas[subset.Name]
+		}
+	}
+
+	return partitions
+}
+
+var subsetReplicasFn = subSetReplicas
+
+func subSetReplicas(subset *subset.Subset) int32 {
+	return subset.Status.Replicas
+}
+
+func calcNextReplicas(nameToSubset map[string]*subset.Subset, allocatedReplicas map[string]int32, nextPartitions map[string]int32, subsetType subSetType) map[string]int32 {
```

Comment:
Can you add comment to explain the logic ?

Reply:
Remote it as partition policy updated.